core/src/main/java/com/linecorp/armeria/common/stream/AsyncMapStreamMessage.java

```diff
@@ -0,0 +1,241 @@
+
+        @Nullable
+        private volatile Subscription upstream;
+        private volatile boolean canceled;
+        private volatile AtomicLong requestedByDownstream;
+
+        private int pendingRequests;
+        private boolean isCompleting;
+
+        AsyncMapSubscriber(Subscriber<? super U> downstream,
+                           Function<T, CompletableFuture<U>> function,
+                           EventExecutor executor) {
+            requireNonNull(downstream, "downstream");
+            requireNonNull(function, "function");
+            requireNonNull(executor, "executor");
+
+            this.downstream = downstream;
+            this.function = function;
+            this.executor = executor;
+            requestedByDownstream = new AtomicLong(0);
+        }
+
+        @Override
+        public void onSubscribe(Subscription subscription) {
+            requireNonNull(subscription, "subscription");
+            upstream = subscription;
+            downstream.onSubscribe(this);
+        }
+
+        @Override
+        public void onNext(T item) {
+            requireNonNull(item, "item");
+
+            if (canceled) {
+                StreamMessageUtil.closeOrAbort(item);
+                return;
+            }
+
+            try {
+                final CompletableFuture<U> future = function.apply(item);
+                requireNonNull(future, "function.apply() returned null");
+
+                pendingRequests++;
+                future.handle((res, cause) -> {
+                    if (executor.inEventLoop()) {
+                        publishDownstream(res, cause);
+                    } else {
+                        executor.execute(() -> publishDownstream(res, cause));
+                    }
+                    return null;
+                });
+            } catch (Throwable ex) {
+                StreamMessageUtil.closeOrAbort(item, ex);
+                upstream.cancel();
+                onError(ex);
+            }
+        }
+
+        private void publishDownstream(@Nullable U item, @Nullable Throwable cause) {
+            if (canceled) {
+                if (item != null) {
+                    StreamMessageUtil.closeOrAbort(item);
+                }
+
+                return;
+            }
+
+            try {
+                if (cause != null) {
+                    upstream.cancel();
+                    onError(cause);
+                } else {
+                    requireNonNull(item, "function.apply()'s future completed with null");
+                    downstream.onNext(item);
+
+                    if (--pendingRequests == 0 && isCompleting) {
+                        downstream.onComplete();
+                    } else if (requestedByDownstream.get() > 0) {
+                        if (requestedByDownstream.get() != Long.MAX_VALUE) {
+                            requestedByDownstream.decrementAndGet();
+                        }
+                        upstream.request(1);
+                    }
```

Comment:
I guess we shouldn't call `upstream.request(1)` when `isCompleting` is `true` and `pendingRequests > 0`
```suggestion
                    if (isCompleting) {
                        if (--pendingRequests == 0) {
                            downstream.onComplete();        
                        }
                    } else if (requestedByDownstream.get() > 0) {
                        if (requestedByDownstream.get() != Long.MAX_VALUE) {
                            requestedByDownstream.decrementAndGet();
                        }
                        upstream.request(1);
                    }
```